### ISSUE TITLE
Feature: in-place bitwise operations

### ISSUE THREAD

**Issue by huguesb** [tower120/hi_sparse_bitset#38]
I'm doing some graph manipulation, with the graph being represented by adjacency list, where each vertex is an integer. Using BitSet as a replacement for `HashSet<usize>` is making things considerably faster, however the inability to do in-place bitwise operations on BitSet is really annoying and forcing me to do a lot of tedious `for` loops to add or remove entries.

**Comment by tower120** [comment]
Can you show example of " a lot of tedious for loops to add or remove entries."?

Also, if you do a graph - could this https://crates.io/crates/hibit_tree be more useful for you? You can use it as a Map<usize, Vert> and can do inter-set operations on it. It is faster then intesrsecting a set, and then accessing a HashMap by index.

This would be a more or less https://docs.rs/hibit_tree/0.1.0-alpha.3/hibit_tree/struct.SparseTree.html equal to `hi_sparse_bitset::BitSet`

**Comment by huguesb** [comment]
> Can you show example of " a lot of tedious for loops to add or remove entries."?

Mostly it's having to turn every `a.extend(b)` into an explicit `for x in b { a.insert(x) }` when I would have expected to be able to just do `a |= b`

> Also, if you do a graph - could this https://crates.io/crates/hibit_tree be more useful for you? You can use it as a Map<usize, Vert> and can do inter-set operations on it. It is faster then intesrsecting a set, and then accessing a HashMap by index.
Neat! But not really relevant to my use case :)

**Comment by tower120** [comment]
Ok - I'll see what I can do.
I may do some refactor as well, so let me ask a few questions: 
- Do you use `BitSet` or `SmallBitSet`?
- What is your index-range? What max index you use within set?
- Do you need compile-time configurable hierarchy depth? The shallower - the faster, the deeper - the higher index-range.

**Comment by huguesb** [comment]
>     * Do you use `BitSet` or `SmallBitSet`?

`BitSet`

>     * What is your index-range? What max index you use within set?

I'm computing dependency graphs from python code so the max index is basically the number of Python modules under consideration, which varies by repository. For the big repo that I'm focusing on right now that's ~70k

>     * Do you need compile-time configurable hierarchy depth? The shallower - the faster, the deeper - the higher index-range.

No. In my case, the index range is only known at run-time.


While I have your attention and you're considering some changes, it would be nice to have some helper functions for serialization. Right now I'm doing the dumbest possible thing: storing each bitset as a length-prefixed sequence of varint, which means I need to iterate once to get the length with`.iter().count()`, and then a second time to get the indices. This makes ser/des unnecessarily slow, and the stored data high overhead (on the plus side, indices are sorted so it's highly compressible). Having an easy way to serialize and deserialize as a sequence of data blocks would be much preferrable. Bonus point if the API is not tied to any specific ser/des framework (fwiw I'm using [speedy](https://docs.rs/speedy/latest/speedy/) )

**Comment by tower120** [comment]
Are you aware of BitSet memory overhead? See memory layout scheme on doc's main page.

> No. In my case, the index range is only known at run-time.

But the BitSet have limited range. It's depth always fixed, for performance. Currently there is 3 levels. And you can configure only block width. I'm talking about configuring it's depth as well. For example, if you would fit u16::MAX range - you could deal with 2level x 256bit configuration - that will both reduce hierarchy memory overhead and speed up all operations.

>  ... length-prefixed sequence of varint ...
 
What is that?

>  need to iterate once to get the length with.iter().count()

There is currently no len counter at all. And it can be only implemented for physical BitSet (I mean not for the lazy-one, like
intersection result). For lazy-ones the fastest theoretical way to get len is by summing len of all data blocks `block_iter().for_each(|block| block.len())`.

If you actually need len in BitSet specifically - yes I can add that. Make an separate issue if that so -  I may forget that.

> Having an easy way to serialize and deserialize as a sequence of data blocks would be much preferrable. 

You CAN iterate over datablocks now https://docs.rs/hi_sparse_bitset/0.6.1/hi_sparse_bitset/#datablocks with https://docs.rs/hi_sparse_bitset/0.6.1/hi_sparse_bitset/trait.BitSetInterface.html#method.block_iter. It returns https://docs.rs/hi_sparse_bitset/0.6.1/hi_sparse_bitset/struct.DataBlock.html .  You should be able to serialize it. The problem is that there is no interface for pushing blocks for efficient deserialization. This I can add.

**Comment by huguesb** [comment]
> Are you aware of BitSet memory overhead? See memory layout scheme on doc's main page.
> 
> > No. In my case, the index range is only known at run-time.
> 
> But the BitSet have limited range. It's depth always fixed, for performance. Currently there is 3 levels. And you can configure only block width. I'm talking about configuring it's depth as well. For example, if you would fit u16::MAX range - you could deal with 2level x 256bit configuration - that will both reduce hierarchy memory overhead and speed up all operations.

Yeah, I went with the `_128` version to be safe. I could probably have a more optimized hierarchy but it was enough of an improvement over `HashSet` that I didn't bother tweaking further. If there was an easy way to pick the right hierarchy based on the max range at run-time I'd use that but I don't want to hard-code a small range at compile time.

> > ... length-prefixed sequence of varint ...
> 
> What is that?

https://protobuf.dev/programming-guides/encoding/#varints


> > need to iterate once to get the length with.iter().count()
> 
> There is currently no len counter at all. And it can be only implemented for physical BitSet (I mean not for the lazy-one, like intersection result). For lazy-ones the fastest theoretical way to get len is by summing len of all data blocks `block_iter().for_each(|block| block.len())`.
> 
> If you actually need len in BitSet specifically - yes I can add that. Make an separate issue if that so - I may forget that.
honestly, the only reason I need the length right now is as an implementation detail for serialization, and sometimes for debug printing so not a huge deal if there is a better serialization option


> > Having an easy way to serialize and deserialize as a sequence of data blocks would be much preferrable.
> 
> You CAN iterate over datablocks now https://docs.rs/hi_sparse_bitset/0.6.1/hi_sparse_bitset/#datablocks with https://docs.rs/hi_sparse_bitset/0.6.1/hi_sparse_bitset/trait.BitSetInterface.html#method.block_iter. It returns https://docs.rs/hi_sparse_bitset/0.6.1/hi_sparse_bitset/struct.DataBlock.html . You should be able to serialize it. The problem is that there is no interface for pushing blocks for efficient deserialization. This I can add.
That would be great!

**Comment by tower120** [comment]
@huguesb FYI, there is highly efficient seralization/deserealization implementation as part of https://github.com/tower120/hi_sparse_bitset/issues/40 now in `dev` branch. There is also `insert_block()` and `BitSet::from(impl BitSetInterface)` - that efficiently materialize lazy bitsets. Better late then never...

**Comment by tower120** [comment]
In-place union implemented in https://github.com/tower120/hi_sparse_bitset/pull/54
through `BitSet::union_with()`, `BitSet::into_union()` and `|=`, `|` operations.

**Comment by tower120** [comment]
@vaivaswatha if you could test in-place union - that would be great! `BitSet::from(impl BitSetInterface)` also was changed.

**Comment by vaivaswatha** [comment]
> [@vaivaswatha](https://github.com/vaivaswatha) if you could test in-place union - that would be great! `BitSet::from(impl BitSetInterface)` also was changed.

I tried this out on my project locally (replacing `reduce(Or, ...)` with a loop and the new `|=`). My testsuite passes fine. Thank you 🙏 @tower120 .

**Comment by tower120** [comment]
@vaivaswatha I wonder if in-place union loop is actually more efficient then `reduce(Or, ...)` though... 
BTW, you can combine both `a |= reduce(Or, ...)`.
Also, there is `a | b` or `union_into(self, other)` - it automatically merge into biggest container and return it.

**Comment by vaivaswatha** [comment]
> @vaivaswatha I wonder if in-place union loop is actually more efficient then `reduce(Or, ...)` though... 
> BTW, you can combine both `a |= reduce(Or, ...)`.
> Also, there is `a | b` or `union_into(self, other)` - it automatically merge into biggest container and return it.

I sort of tried this as a correctness test, assuming that `reduce()` would be more efficient. So it was just a local change. Are you suggesting that `|=` in a loop may by more efficient?

I do have places where it'll just be a single `|=` (not in  loop) and this will be convenient to use. I haven't gotten to that yet.

**Comment by tower120** [comment]
> I sort of tried this as a correctness test, assuming that reduce() would be more efficient. So it was just a local change. Are you suggesting that |= in a loop may by more efficient?

No, I thought that you replaced `reduce()` with `|=` loop for efficiency. I actually don't know... I feel that `a |= reduce(Or, ...)` should be the most efficient form, but I don't know. All versions should be quite efficient.

--- 

I must say, there is one design compromise - each hierarchy block have indices array. And that indices array filled with 0's on construction. This design allows branchless `contains()`, and maybe speed ups something else (I don't remember).
Those hierarchy blocks reused after becoming empty and removed.  When reused they are not constructed again.
As an alternative design - that array could be left uninitialized on construction, but `contains()` will have 2 ifs.. On my benchmarks that design cut performance of `contains` almost half. It was initially designed around lazy bitsets - since I needed complex intersection lists - there were almost no `materialization`, even if I need to "test bit" - I used virtual set directly.

But I guess theoretically, I can shift towards design with more lightweight construction... But I don't know if it actually worth it - 128bit indices array is 256 bytes - I can't say if that would be noticeable.

**Comment by vaivaswatha** [comment]
> No, I thought that you replaced reduce() with |= loop for efficiency. I actually don't know... I feel that a |= reduce(Or, ...) should be the most efficient form, but I don't know. All versions should be quite efficient.

For now, my `master` continues to use `reduce` because otherwise I'll need `|=` in a loop. But I do have upcoming changes where `|=` will be more convenient. So overall, this addition is definitely beneficial. Thank you.

**Comment by tower120** [comment]
In-place intersection operations added in https://github.com/tower120/hi_sparse_bitset/pull/55.

**Comment by tower120** [comment]
In-place intersection and union released in v0.7.5.

I'm closing this now. If someone needs other in-place operations reopen it or make a new issue-feature-request.
